File: backend/app/services/pet_service.py
from app.repositories.pet_repository import PetRepository
from app.repositories.user_repository import UserRepository


class PetService:
    """Serviço de gerenciamento de pets"""
    
    @staticmethod
    def list_available(filters=None):
        """Lista pets disponíveis com filtros opcionais"""
        return PetRepository.find_all(filters)
    
    @staticmethod
    def get_pet(pet_id):
        """Obtém detalhes de um pet"""
        pet = PetRepository.find_by_id(pet_id)
        if not pet:
            return None, 'Pet não encontrado'
        return pet, None
    
    @staticmethod
    def create_pet(owner_id, pet_data):
        """Cria novo pet (apenas ONG)"""
        owner = UserRepository.find_by_id(owner_id)
        if not owner:
            return None, 'Usuário não encontrado'
        
        if owner.tipo != 'ONG':
            return None, 'Apenas ONGs podem cadastrar pets'
        
        # Validações básicas
        if not pet_data.get('nome'):
            return None, 'Nome do pet é obrigatório'
        if not pet_data.get('especie'):
            return None, 'Espécie é obrigatória'
        if not pet_data.get('whatsapp_contato'):
            return None, 'WhatsApp de contato é obrigatório'
        
        pet_create_data = {
            'owner_id': owner_id,
            'nome': pet_data['nome'],
            'especie': pet_data['especie'],
            'raca': pet_data.get('raca'),
            'idade': pet_data.get('idade'),
            'sexo': pet_data.get('sexo'),
            'porte': pet_data.get('porte'),
            'castrado': pet_data.get('castrado', False),
            'vacinado': pet_data.get('vacinado', False),
            'localizacao': pet_data.get('localizacao'),
            'descricao': pet_data.get('descricao'),
            'imagem_url': pet_data.get('imagem_url'),
            'whatsapp_contato': pet_data['whatsapp_contato'],
            'status': 'DISPONIVEL'
        }
        
        pet = PetRepository.create(pet_create_data)
        return pet, None
    
    @staticmethod
    def update_pet(user_id, pet_id, update_data):
        """Atualiza dados do pet (apenas dono da ONG)"""
        pet = PetRepository.find_by_id(pet_id)
        if not pet:
            return None, 'Pet não encontrado'
        
        if pet.owner_id != user_id:
            return None, 'Você não tem permissão para editar este pet'
        
        # Permitir atualização de certos campos
        allowed_fields = ['raca', 'idade', 'sexo', 'porte', 'castrado', 'vacinado', 
                         'localizacao', 'descricao', 'imagem_url', 'whatsapp_contato']
        filtered_data = {k: v for k, v in update_data.items() if k in allowed_fields}
        
        pet = PetRepository.update(pet_id, filtered_data)
        return pet, None
    
    @staticmethod
    def delete_pet(user_id, pet_id):
        """Deleta um pet (apenas dono da ONG)"""
        pet = PetRepository.find_by_id(pet_id)
        if not pet:
            return None, 'Pet não encontrado'
        
        if pet.owner_id != user_id:
            return None, 'Você não tem permissão para deletar este pet'
        
        success = PetRepository.delete(pet_id)
        return success, None
    
    @staticmethod
    def update_status(user_id, pet_id, new_status):
        """Atualiza status do pet (apenas dono da ONG)"""
        pet = PetRepository.find_by_id(pet_id)
        if not pet:
            return None, 'Pet não encontrado'
        
        if pet.owner_id != user_id:
            return None, 'Você não tem permissão para alterar o status deste pet'
        
        valid_statuses = ['DISPONIVEL', 'EM_PROCESSO', 'ADOTADO', 'INATIVO']
        if new_status not in valid_statuses:
            return None, f'Status inválido. Valores aceitos: {", ".join(valid_statuses)}'
        
        pet = PetRepository.update_status(pet_id, new_status)
        return pet, None
    
    @staticmethod
    def list_user_pets(user_id):
        """Lista pets de um usuário (apenas ONG)"""
        user = UserRepository.find_by_id(user_id)
        if not user:
            return None, 'Usuário não encontrado'
        
        if user.tipo != 'ONG':
            return None, 'Apenas ONGs têm pets'
        
        pets = PetRepository.find_by_owner(user_id)
        return pets, None

File: frontend/lib/services/pet_service.dart
import 'package:shared_preferences/shared_preferences.dart';

import '../config/api_config.dart';
import '../models/pet.dart';
import 'http_helper.dart';

class PetService {
  /// Obtém token do armazenamento
  Future<String?> _token() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('token');
  }

  /// Cria novo pet (apenas ONG)
  Future<void> createPet({
    required String nome,
    required String especie,
    required String raca,
    required String idade,
    required String sexo,
    required String porte,
    required bool castrado,
    required bool vacinado,
    required String localizacao,
    required String descricao,
    required String imagemUrl,
    required String whatsapp,
  }) async {
    try {
      final token = await _token();
      if (token == null) {
        throw HttpException('Token não encontrado. Faça login novamente.');
      }

      await HttpHelper.post(
        '${ApiConfig.baseUrl}/pets',
        body: {
          'nome': nome,
          'especie': especie,
          'raca': raca,
          'idade': idade,
          'sexo': sexo,
          'porte': porte,
          'castrado': castrado,
          'vacinado': vacinado,
          'localizacao': localizacao,
          'descricao': descricao,
          'imagem_url': imagemUrl,
          'whatsapp_contato': whatsapp,
        },
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      print('✅ Pet cadastrado com sucesso');
    } catch (e) {
      print('❌ Erro ao cadastrar pet: $e');
      rethrow;
    }
  }

  /// Lista todos os pets disponíveis
  Future<List<Pet>> getPets({
    String? sexo,
    String? porte,
    String? especie,
  }) async {
    try {
      String url = '${ApiConfig.baseUrl}/pets';
      
      // Adiciona filtros se fornecidos
      final queryParams = <String>[];
      if (sexo != null && sexo.isNotEmpty) queryParams.add('sexo=$sexo');
      if (porte != null && porte.isNotEmpty) queryParams.add('porte=$porte');
      if (especie != null && especie.isNotEmpty) queryParams.add('especie=$especie');
      
      if (queryParams.isNotEmpty) {
        url = '$url?${queryParams.join('&')}';
      }

      final response = await HttpHelper.get(url);

      // Trata resposta como lista
      final List<dynamic> jsonList;
      if (response is List) {
        jsonList = response;
      } else if (response is Map && response.containsKey('pets')) {
        jsonList = response['pets'] ?? [];
      } else {
        throw Exception('Formato de resposta inesperado');
      }

      print('✅ Carregados ${jsonList.length} pets');
      return jsonList.map((e) => Pet.fromJson(e)).toList();
    } catch (e) {
      print('❌ Erro ao buscar pets: $e');
      rethrow;
    }
  }

  /// Obtém detalhes de um pet específico
  Future<Pet> getPet(int id) async {
    try {
      final response = await HttpHelper.get(
        '${ApiConfig.baseUrl}/pets/$id',
      );

      print('✅ Pet $id carregado');
      return Pet.fromJson(response);
    } catch (e) {
      print('❌ Erro ao buscar pet $id: $e');
      rethrow;
    }
  }

  /// Lista pets do usuário logado (ONG)
  Future<List<Pet>> getMyPets() async {
    try {
      final token = await _token();
      if (token == null) {
        throw HttpException('Token não encontrado. Faça login novamente.');
      }

      final response = await HttpHelper.get(
        '${ApiConfig.baseUrl}/users/me/pets',
        headers: {
          'Authorization': 'Bearer $token',
        },
      );

      // Trata resposta como lista
      final List<dynamic> jsonList;
      if (response is List) {
        jsonList = response;
      } else if (response is Map && response.containsKey('pets')) {
        jsonList = response['pets'] ?? [];
      } else {
        throw Exception('Formato de resposta inesperado');
      }

      print('✅ Carregados ${jsonList.length} pets próprios');
      return jsonList.map((e) => Pet.fromJson(e)).toList();
    } catch (e) {
      print('❌ Erro ao buscar meus pets: $e');
      rethrow;
    }
  }

  /// Atualiza informações de um pet
  Future<void> updatePet(
    int id, {
    required String nome,
    required String especie,
    required String raca,
    required String idade,
    required String sexo,
    required String porte,
    required bool castrado,
    required bool vacinado,
    required String localizacao,
    required String descricao,
    required String imagemUrl,
    required String whatsapp,
  }) async {
    try {
      final token = await _token();
      if (token == null) {
        throw HttpException('Token não encontrado. Faça login novamente.');
      }

      await HttpHelper.put(
        '${ApiConfig.baseUrl}/pets/$id',
        body: {
          'nome': nome,
          'especie': especie,
          'raca': raca,
          'idade': idade,
          'sexo': sexo,
          'porte': porte,
          'castrado': castrado,
          'vacinado': vacinado,
          'localizacao': localizacao,
          'descricao': descricao,
          'imagem_url': imagemUrl,
          'whatsapp_contato': whatsapp,
        },
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      print('✅ Pet $id atualizado');
    } catch (e) {
      print('❌ Erro ao atualizar pet $id: $e');
      rethrow;
    }
  }

  /// Deleta um pet
  Future<void> deletePet(int id) async {
    try {
      final token = await _token();
      if (token == null) {
        throw HttpException('Token não encontrado. Faça login novamente.');
      }

      await HttpHelper.delete(
        '${ApiConfig.baseUrl}/pets/$id',
        headers: {
          'Authorization': 'Bearer $token',
        },
      );

      print('✅ Pet $id deletado');
    } catch (e) {
      print('❌ Erro ao deletar pet $id: $e');
      rethrow;
    }
  }

  /// Altera status de um pet
  Future<void> updatePetStatus(int id, String status) async {
    try {
      final token = await _token();
      if (token == null) {
        throw HttpException('Token não encontrado. Faça login novamente.');
      }

      await HttpHelper.put(
        '${ApiConfig.baseUrl}/pets/$id/status',
        body: {
          'status': status,
        },
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      print('✅ Status do pet $id atualizado para $status');
    } catch (e) {
      print('❌ Erro ao atualizar status do pet $id: $e');
      rethrow;
    }
  }
}